subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -41,9 +43,9 @@
     private final Map<String, ImmutableList<Path>> watchedRootsForSnapshot = new HashMap<>();
     private final Multiset<Path> shouldWatchDirectories = HashMultiset.create();
```

Comment:
Why is this a `MultiSet` instead of a `Set`?

Reply:
This is a `MultiSet` since the same directory can be watched because of two different snapshots. So we need to continue watching a directory even though one of the two references has been removed.

The easiest example are two `RegularFileSnapshot`s in the same directory. We need to watch the directory for the two file snapshots, and keep watching it even if one of the two snapshots is removed.